core/src/consensus.rs

```diff
@@ -108,6 +109,12 @@ pub struct Tower {
     threshold_size: f64,
     lockouts: VoteState,
     last_vote: Vote,
+    #[serde(skip)]
+    // The blockhash used in the last vote transaction, may or may not equal the
+    // blockhash of the voted block itself, depending if the vote slot was refreshed.
+    // For instance, a vote for slot 5, may be refreshed/resubmitted for inclusion in
+    //  block 10, in  which case `last_vote_tx_blockhash` equals the blockhash of 10, not 5.
+    last_vote_tx_blockhash: Hash,
```

Comment:
Use an `Option<Hash>` here otherwise a node that just loaded their tower from disk will think `Hash::default()` isn't in the recent blockhash queue and retry their latest recorded vote even if it was recently sent

Reply:
Also, does this need to be stored here? Could it be recovered from gossip instead?

Reply:
I think recovering from gossip will be a lot more complexity

It should be fine for them to retry their latest recorded vote, as it should only happen *once* on startup. Also, if the vote was pushed by this node and already exists in gossip, this check here: https://github.com/solana-labs/solana/pull/16735/files#diff-acba1816adaf11741b4cd19f39d543dbdb6a1af4726544a8b61a6c4db616c76cR1215-R1219 will ensure the node does not replace that vote in gossip before receiving it from the network first.

Reply:
> will ensure the node does not replace [ ... ]

Doesn't that ensure the node *does* replace the vote in gossip?

Reply:
So if the vote doesn't already exist, `vote_index` will be `None`, so nothing new will be pushed